src/nvim/tui/tui.c

```diff
@@ -1240,11 +1260,31 @@ static void out(void *ctx, const char *str, size_t len)
   size_t available = sizeof(data->buf) - data->bufpos;
 
   if (len > available) {
-    flush_buf(ui, false);
+    flush_buf(ui);
+  }
+
+  memcpy(data->buf + data->bufpos, str, len);
+  data->bufpos += len;
+}
+
+static void limited_out(void *ctx, const char *str, size_t len)
```

Comment:
Any reason not to replace `out()` with this entirely? And maybe store the state on `TUIData` instead of the new `limited_out_ctx` struct.

Reply:
I didn't want to modify the current users of `out()`, both because this slightly changes the semantics (possibly causing flushes at different points) and requires them to pass a context struct.

The semantics part shouldn't be an issue though, so one could reimplement `out()` using `limited_out()` which would reduce duplicated code. I'll do that.

Rolling `llimited_out_ctx` into `TUIData` feels slightly wrong because this is local state, but I have no strong opinion on this. If you think this is better stored as part of `TUIData`, I'll move it there. your call.

Reply:
> The semantics part shouldn't be an issue though, so one could reimplement out() using limited_out() which would reduce duplicated code. I'll do that.

Ok, will wait for that before merging.

> Rolling llimited_out_ctx into TUIData feels slightly wrong because this is local state, but I have no strong opinion on this. If you think this is better stored as part of TUIData, I'll move it there. Your call.

Let's leave it as you have it. TUIData has fields like `bufpos`, but what you have is probably clearer and otherwise harmless.